docs: add normalized LM API migration plan (#9828)
stanfordnlp/dspy: docs/docs/community/normalized-lm-api-migration.md

# Typed LM API migration plan

DSPy is moving toward a typed language-model boundary while keeping `dspy.BaseLM` as the public base class for language models.

> The short version: **most DSPy users do not need to change anything in DSPy 3.3**. Existing `lm(...)`, modules, and programs keep their current behavior by default. The typed LM API is opt-in in 3.3 with `dspy.context(experimental=True)`.

Reviewer: ```suggestion
**Most DSPy users do not need to change anything in DSPy 3.3**. 

Existing `lm(...)`, modules, and programs keep their current behavior by default. The typed LM API is opt-in in 3.3 with `dspy.context(experimental=True)`.

TLDR: dspy.LM.forward currently is untyped, and has scattered half LiteLLM half custom to work around liteLLM logic.

We will be migrating `BaseLM.forward` and `LM.forward` from:
```python
def forward(self, prompt=None, messages=None, **kwargs):
```
to
```python
(def forward(self, request: dspy.LMRequest) -> dspy.LMResponse)```
```
(git formatting ruins the nested blocks)

!!! note "Status"
    This is a migration plan for the DSPy 3.3–3.6/4.0 series. Names and exact release timing may change before implementation lands, but the staged compatibility plan below should guide discussion.

!!! info "Community feedback wanted"
    This plan mostly affects custom LM backends and adapters. If you maintain one, please review the proposed one-line `forward_contract` migration and share feedback before the default LM path changes.

## Who is affected?

> | Group | DSPy 3.3 action | Later action |

Reviewer: Lets say 3.4+

| --- | --- | --- |
| Most DSPy users | No action required. | Optionally try typed direct LM calls with `dspy.context(experimental=True)`. |
| Custom LM authors | Add `forward_contract = "legacy"` if your LM implements `forward(prompt=None, messages=None, **kwargs)`. | Migrate to `forward_contract = "typed_lm"` and `forward(request: dspy.LMRequest) -> dspy.LMResponse` before legacy support is removed. |

> | New LM backend authors | Prefer the typed contract from the start. | Package and distribute backends as first-class DSPy-compatible libraries. |

Reviewer: "Backend" implies that this is different than LM

| Custom adapter authors | Do not call `lm.forward(...)` directly. Route LM calls through `lm(...)`. | Move toward building `LMRequest` objects and parsing `LMResponse` directly. |
| DSPy maintainers and contributors | Keep public typed returns guarded by `experimental=True` in 3.3. | Warn in 3.4, make the typed path the default candidate in 3.5, and remove legacy in 3.6 or 4.0 after final review. |

## Background

> Today, most `BaseLM` subclasses implement a legacy provider-shaped hook:

Reviewer: ```suggestion
Today, `BaseLM` subclasses implement an untyped forward method, with a few optional parameters:
```

```python
def forward(self, prompt=None, messages=None, **kwargs):
    ...
```

> That hook usually receives OpenAI/LiteLLM-shaped inputs and returns an OpenAI-like provider response. DSPy then post-processes that response into legacy outputs such as `list[str | dict]`.

Reviewer: ```suggestion
That hook usually receives OpenAI/LiteLLM-shaped inputs and returns an OpenAI-like provider response. DSPy then post-processes that response into a `list[str | dict]` containing outputs.
```

> The target contract is typed:

Reviewer: ```suggestion
Because the current parameters are untyped, it is hard to know inside of an LM exactly which inputs you will get and what types they will contain. Our new contract is typed and provider neutral. We have designed the `LMRequest` and `LMResponse` objects to be flexible enough such that any LLM Provider's API easily supports this:
```

```python
def forward(self, request: dspy.LMRequest) -> dspy.LMResponse:
    ...
```

DSPy has settled on the internal LM type system around `LMRequest`, `LMResponse`, typed messages, parts, config, usage, and stream events. These types should be treated as the stable direction for LM backends. Concrete LMs translate between these DSPy types and their provider API.

## Why this matters

> The typed boundary gives DSPy one clear internal representation for LM calls:

Reviewer: This also clarifies the job of an adapter, which is to translate from signature + inputs -> LMRequest.

```python
class EchoLM(dspy.BaseLM):
    forward_contract = "typed_lm"

    def forward(self, request: dspy.LMRequest) -> dspy.LMResponse:
        return dspy.LMResponse.from_text("hello", model=request.model)
```

## Guide for custom adapter authors

Adapters should call the LM object, not `forward()` directly.

Preferred typed boundary:

```python
request = dspy.LMRequest.from_call(
    model=lm.model,
    messages=messages,
    **lm_kwargs,
)
response = lm(request)
```

Avoid this in adapters:

```python
lm.forward(...)
```

`BaseLM.__call__()` is the compatibility boundary. It owns input normalization, choosing the legacy or typed `forward()` path, adapting legacy outputs into `LMResponse`, and preserving public return behavior unless `experimental=True` is enabled.

During the transition, adapters may still convert `LMResponse` back to legacy parser inputs. The long-term direction is for adapters to parse `LMResponse` directly.

## Version sequence

| Version | Missing `forward_contract` | Public `lm(...)` behavior |
| --- | --- | --- |
| 3.3 | Treat as legacy | Typed returns available only through `experimental=True` or explicit `LMRequest` calls. |

> | 3.4 | Treat as legacy and warn | Still guarded while migration continues. |

Reviewer: ```suggestion
| 3.4 | Treat as legacy and warn | Still require `experimental=True` or explicit `LMRequest` while migration continues. |
```

> | 3.5 | Require explicit contract or flip default after final review | Candidate default typed path. |

Reviewer: ```suggestion
| 3.5 | Require explicit contract or flip default after final review | Typed path become default with legacy escape hatch. |
```

| 3.6 or 4.0 | Remove the legacy `forward(prompt, messages, **kwargs)` contract after final review | `forward(request: LMRequest) -> LMResponse` is the only supported `BaseLM` implementation contract. |

> Before changing the default, DSPy should give custom LM authors enough time to add one of:

Reviewer: ```suggestion
Before changing the default, DSPy will give custom LM authors enough time to add one of:
```
